lib/srv/alpnproxy/aws_local_proxy.go

```diff
@@ -46,11 +56,168 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
+// │                         ├───────────────────►│ restore headers       │
+// │                         │                    ├──────────────────────►│
+// │                         │                    │                       │
+// │                         │                    │◄──────────────────────┤
+// │                         │◄───────────────────┤                       │
+// │◄────────────────────────┤                    │                       │
+//
+// Note that the first sts:AssumeRole should be signed with the
+// local-proxy-generated credentials by the AWS client, while the second
+// request is signed with real credentials of the assumed role.
 func (m *AWSAccessMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
-	if err := aws.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
+	sigV4, err := awsutils.ParseSigV4(req.Header.Get(awsutils.AuthorizationHeader))
+	if err != nil {
+		m.Log.WithError(err).Error("Failed to parse AWS request authorization header.")
+		rw.WriteHeader(http.StatusForbidden)
+		return true
+	}
+
+	if assumedRole, found := m.findAssumedRole(sigV4.KeyID); found {
+		return m.handleRequestByAssumedRole(rw, req, assumedRole)
+	}
+	return m.handleCommonRequest(rw, req)
+}
+
+func (m *AWSAccessMiddleware) handleCommonRequest(rw http.ResponseWriter, req *http.Request) bool {
+	if err := awsutils.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
 		m.Log.WithError(err).Error("AWS signature verification failed.")
 		rw.WriteHeader(http.StatusForbidden)
 		return true
 	}
 	return false
 }
+
+func (m *AWSAccessMiddleware) handleRequestByAssumedRole(rw http.ResponseWriter, req *http.Request, assumedRole *sts.AssumeRoleOutput) bool {
+	credentials := credentials.NewStaticCredentials(
+		aws.ToString(assumedRole.Credentials.AccessKeyId),
+		aws.ToString(assumedRole.Credentials.SecretAccessKey),
+		aws.ToString(assumedRole.Credentials.SessionToken),
+	)
+
+	if err := awsutils.VerifyAWSSignature(req, credentials); err != nil {
+		m.Log.WithError(err).Error("AWS signature verification failed.")
+		rw.WriteHeader(http.StatusForbidden)
+		return true
+	}
+
+	m.Log.Debugf("Rewriting headers for AWS request by assumed role %q.", aws.ToString(assumedRole.AssumedRoleUser.Arn))
+
+	// Add a custom header for marking the special request.
+	req.Header.Add(appcommon.TeleportAWSAssumedRole, aws.ToString(assumedRole.AssumedRoleUser.Arn))
+
+	// Rename the original authorization header to ensure older app agents
+	// (that don't support the requests by assumed roles) will fail.
+	utils.RenameHeader(req.Header, awsutils.AuthorizationHeader, appcommon.TeleportAWSAssumedRoleAuthorization)
+	return false
+}
+
+func (m *AWSAccessMiddleware) HandleResponse(response *http.Response) error {
+	if response == nil || response.Request == nil {
+		return nil
+	}
+
+	authHeader := utils.GetAnyHeader(response.Request.Header,
+		awsutils.AuthorizationHeader,
+		appcommon.TeleportAWSAssumedRoleAuthorization,
```

Comment:
Why check for `TeleportAWSAssumedRoleAuthorization` here? Won't it always be rewritten to the authorization header?

Reply:
> Won't it always be rewritten to the authorization header?

This only happens on the server side to AWS. The `HandleResonse` for requests by assumed roles will see the `appcommon.TeleportAWSAssumedRoleAuthorization` header. Just double-checking here in case its "nested" assume role calls.